clippy_lints/src/methods/mod.rs

```diff
@@ -4253,6 +4254,36 @@ declare_clippy_lint! {
     "map of a trivial closure (not dependent on parameter) over a range"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Detects when an entire collection is being cloned eagerly, instead of each item lazily.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Cloning a collection requires allocating space for all elements and cloning each element into this new space,
+    /// whereas using `Iterator::cloned` does not allocate any more space and only requires cloning each element as they are consumed.
```

Comment:
I'm guessing this highly depends on the iterator usage. For a vector, it makes a difference of allocating a big chunk of memory once against several small allocations. For vectors it might actually be the other way around (meaning `.clone().into_iter()` being faster), but this is just guesstimating, since I haven't benchmarked this. 

One thing that's safe to say, is that cloning each item individually will reduce the memory footprint, but only if the items are dropped and not stored somewhere afterward.

Could you try to benchmark this?

I've just checked the issue, and that one was suggestion to lint on `cloned().into_iter().filter()`. Linting on that should be safe again